### ISSUE TITLE
Render asset as a collision mesh when container is not loaded

### ISSUE THREAD

**Issue by LeXXik** [playcanvas/engine#3638]
We have a framework that loads a level, stored as a template. That level contains the level meshes, imported as a hierarchy (amazing feature, by the way, cudos!). They are then rigged with collision and rigidbody components, scripts, etc. Each level is loaded and instantiated via script.

On one of the levels I tried to add a piece that is a dynamic body with a render asset as a mesh collision. However, the app crashes, as soon as the level is loaded:

![image](https://user-images.githubusercontent.com/5677782/139555388-8715f95b-5294-46c9-b267-7b691a84e948.png)

![image](https://user-images.githubusercontent.com/5677782/139555427-68c49ca0-2a5a-4970-9786-f371aab81bba.png)

I tried to isolate it and create a blank project with that piece only in a template. However, the template is loaded fine and `_meshes` property is not null. I tried both instantiating the template via script and just add it in Editor. 

I am a bit stuck, and looking for advice on the conditions when `_meshes` property could be `null`, while `renderAsset` is not and having a correct reference id. After all the shape is recreated when the asset is loaded:

![image](https://user-images.githubusercontent.com/5677782/139555640-01800ca5-a5bd-4fef-8297-0d0e90f9a999.png)
![image](https://user-images.githubusercontent.com/5677782/139555556-82a9764c-d35b-439d-8e62-a4a3063ca135.png)

**Comment by yaustar** [comment]
Dumb question from me: is the container asset that the render assets are referring to loaded yet?

**Comment by LeXXik** [comment]
It's not a dumb question. This is what the issue is indicating. However, I thought that the shape is recreated on asset loaded callback. If the asset is loaded, then I assume mesh is available? I will add an explicit load to test it.

**Comment by yaustar** [comment]
At a guess without investigating, the render asset is just a bit JSON that references the GLB container asset. So the load completion of the render asset is not dependent on waiting for the GLB container. Much like the sprite and texture atlas relationship.

**Comment by LeXXik** [comment]
I did make sure now the framework loads the asset first. Still errors out. Actually this step is not needed, since collision system does load it automatically, if it finds a render asset is used:
![image](https://user-images.githubusercontent.com/5677782/139588481-f068c221-1403-4be4-af63-e01fbdb9e338.png)

However, later on, in `asset.ready` the callback is returned as soon as `asset.resource` property is found. Regardless of what that property contains:

![image](https://user-images.githubusercontent.com/5677782/139588428-9c658bff-cf22-4183-97a7-21ccb1dc7b10.png)

**Comment by LeXXik** [comment]
I was able to solve the issue by patching the `pc.Asset` to check not only for the `resource` property, but also for a defined `meshes` within it:

```js
    if (! this.resource || (this.resource && (this.resource.meshes !== undefined && ! this.resource.meshes))) {
        this.once("load", function (asset) {
            callback.call(scope, asset);
        });        
    } else {
        callback.call(scope, this);
    }
```

I didn't make a PR, since you might have a more elegant solution. Considering you are introducing new handlers now, this doesn't count for them or their custom properties. Perhaps, some more generalized approached should be used. If you want me to PR this patch, let me know.

**Comment by mvaligursky** [comment]
@vkalpias - any thoughts on this?

**Comment by mvaligursky** [comment]
@LeXXik - you workaround unfortunately does not really solve it .. it only works if the container has been loaded already I believe? Either way, it's not a place to introduce a change like this, as you mentioned as well.

The way the render component (and also particle system component, which can render using meshes from render asset) solves this is here (RenderComponent in this case):

<img width="510" alt="Screenshot 2022-02-22 at 15 46 39" src="https://user-images.githubusercontent.com/59932779/155167959-a10f78bf-7450-4cf5-be57-3a3a337873d4.png">

Basically, when the render asset is loaded, and contains meshes, those are used directly (in `_onSetMeshes`).
At the same time `set:meshes` event is being listened to .. which is triggered when container is loaded later and sets the meshes to render asset .

But I don't see any nice way to handle it this way for the collision component .. the use of meshes is lots more complicated.

**Comment by LeXXik** [comment]
@mvaligursky hmm, yeah, that may well be. I guess that patch worked, since it was applied in a framework, that does preload stuff before the collision component is enabled. It could be the container is loaded by then.

**Comment by fabiensabatie** [comment]
Any news on this ? I'm trying to load entities with a collision component using a render asset with preload set to false.
The launcher seems to call createPhysicalShape before the render asset is loaded, which is understandable, but iIs there any way to trigger this function after the asset has loaded ?

**Comment by yaustar** [comment]
> Any news on this ?

We have it in our bug blitz sprint and will be prioritised against the other issues in the list.

Currently the recommended workaround/practise is to load the template assets first before instantiating the template.
